alphadia/transferlearning/train.py

```diff
@@ -896,3 +897,166 @@ def finetune_charge(self, psm_df: pd.DataFrame) -> pd.DataFrame:
         metrics = test_metric_manager.get_stats()
 
         return metrics
+
+    def _test_ccs(
+        self,
+        epoch: int,
+        epoch_loss: float,
+        test_df: pd.DataFrame,
+        metric_accumulator: MetricManager,
+        data_split: str,
+    ) -> bool:
+        """
+        Test the CCS model using the PSM dataframe and accumulate both the training loss and test metrics.
+
+        Parameters
+        ----------
+        epoch : int
+            The current epoch number.
+        epoch_loss : float
+            The train loss value of the current epoch.
+        test_df : pd.DataFrame
+            The PSM dataframe.
+        metric_accumulator : MetricManager
+            The metric manager object.
+        data_split : str
+            The dataset label to test on. e.g. "validation", "train"
+        Returns
+        -------
+        bool
+            Whether to continue training or not based on the early stopping criteria.
+        """
+        continue_training = True
+        if epoch % self.settings["test_interval"] == 0:
+            self.ccs_model.model.eval()
+
+            pred = self.ccs_model.predict(test_df)
+
+            test_input = {
+                "predicted": pred["ccs_pred"].values,
+                "target": test_df["ccs"].values,
+            }
+            val_metrics = metric_accumulator.calculate_test_metric(
+                test_input, epoch, data_split=data_split, property_name="ccs"
+            )
+            if epoch != -1 and data_split == "validation":
+                metric_accumulator.accumulate_metrics(
+                    epoch,
+                    metric=epoch_loss,
+                    metric_name="l1_loss",
+                    data_split="train",
+                    property_name="ccs",
+                )
+                current_lr = self.ccs_model.optimizer.param_groups[0]["lr"]
+                metric_accumulator.accumulate_metrics(
+                    epoch,
+                    metric=current_lr,
+                    metric_name="lr",
+                    data_split="train",
+                    property_name="ccs",
+                )
+                val_loss = val_metrics[val_metrics["metric_name"] == "l1_loss"][
+                    "value"
+                ].values[0]
+                continue_training = self.early_stopping.step(val_loss)
+                logger.progress(
+                    f" Epoch {epoch:<3} Lr: {current_lr:.5f}   Training loss: {epoch_loss:.4f}   validation loss: {val_loss:.4f}"
+                )
+            else:
+                logger.progress(
+                    f" CCS model tested on {data_split} dataset with the following metrics:"
+                )
+                for i in range(len(val_metrics)):
+                    logger.progress(
+                        f" {val_metrics['metric_name'].values[i]:<30}: {val_metrics['value'].values[i]:.4f}"
+                    )
+
+            self.ccs_model.model.train()
```

Comment:
from the name and docstring of this method I would assume this method just tests, but it also trains?

Reply:
thanks for the clarification!